File: sdk/dotnet/DevLogBus.Sdk/DevLogBusClient.cs
namespace DanSherwin.DevLogBus;

using System.Net.Http;
using System.Text;
using System.Text.Json;

public delegate bool DevLogBusRecordFilter(DevLogBusRecord record);

public delegate DevLogBusRecord DevLogBusRecordRedactor(DevLogBusRecord record);

public sealed class DevLogBusClientOptions
{
    public string Endpoint { get; init; } = DevLogBusClient.DefaultHttpEndpoint;
    public string Source { get; init; } = "";
    public TimeSpan Timeout { get; init; } = TimeSpan.FromSeconds(2);
    public DevLogBusRecordFilter? Filter { get; init; }
    public DevLogBusRecordRedactor? Redactor { get; init; }
    public HttpClient? HttpClient { get; init; }
}

public sealed record DevLogBusPublishResult(int Published, bool Filtered);

public sealed class DevLogBusClient : IDisposable
{
    public const string DefaultHttpEndpoint = "http://127.0.0.1:7423";
    public const string RedactedValue = "[REDACTED]";

    private readonly string endpoint;
    private readonly string source;
    private readonly DevLogBusRecordFilter? filter;
    private readonly DevLogBusRecordRedactor? redactor;
    private readonly HttpClient httpClient;
    private readonly bool ownsHttpClient;

    public DevLogBusClient(DevLogBusClientOptions? options = null)
    {
        options ??= new DevLogBusClientOptions();
        endpoint = TrimEndpoint(options.Endpoint);
        source = options.Source ?? "";
        filter = options.Filter;
        redactor = options.Redactor;

        if (options.HttpClient is null)
        {
            var timeout = options.Timeout > TimeSpan.Zero ? options.Timeout : TimeSpan.FromSeconds(2);
            httpClient = new HttpClient { Timeout = timeout };
            ownsHttpClient = true;
        }
        else
        {
            httpClient = options.HttpClient;
        }
    }

    public string Endpoint => endpoint;

    public string Source => source;

    public Task<DevLogBusPublishResult> PublishMessageAsync(
        string level,
        string message,
        IReadOnlyDictionary<string, object?>? attrs = null,
        CancellationToken cancellationToken = default)
    {
        return PublishAsync(new DevLogBusRecord
        {
            Source = source,
            Level = level,
            Message = message,
            Attrs = attrs,
        }, cancellationToken);
    }

    public Task<DevLogBusPublishResult> PublishRawAttrsAsync(
        string level,
        string message,
        string attrsJson,
        CancellationToken cancellationToken = default)
    {
        return PublishAsync(new DevLogBusRecord
        {
            Source = source,
            Level = level,
            Message = message,
            AttrsJson = attrsJson,
        }, cancellationToken);
    }

    public async Task<DevLogBusPublishResult> PublishAsync(
        DevLogBusRecord input,
        CancellationToken cancellationToken = default)
    {
        var record = string.IsNullOrWhiteSpace(input.Source) && !string.IsNullOrWhiteSpace(source)
            ? input with { Source = source }
            : input;

        if (filter is not null && !filter(record))
        {
            return new DevLogBusPublishResult(0, true);
        }
        if (redactor is not null)
        {
            record = redactor(record);
        }

        using var content = new StringContent(record.ToJson(), Encoding.UTF8, "application/json");
        using var response = await httpClient.PostAsync($"{endpoint}/api/records", content, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            throw new HttpRequestException($"DevLogBus publish failed: HTTP {(int)response.StatusCode}");
        }
        return await ReadResultAsync(response, cancellationToken).ConfigureAwait(false);
    }

    public DevLogBusLogger Logger(string? source = null)
    {
        return new DevLogBusLogger(this, string.IsNullOrWhiteSpace(source) ? this.source : source);
    }

    public static DevLogBusRecordFilter DropSources(params string[] sources)
    {
        var blocked = sources
            .Where(source => !string.IsNullOrWhiteSpace(source))
            .Select(source => source.Trim())
            .ToHashSet(StringComparer.Ordinal);
        return record => !blocked.Contains(record.Source);
    }

    public static DevLogBusRecordRedactor RedactMessage()
    {
        return record => record with { Message = RedactedValue };
    }

    public void Dispose()
    {
        if (ownsHttpClient)
        {
            httpClient.Dispose();
        }
    }

    private static async Task<DevLogBusPublishResult> ReadResultAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        if (string.IsNullOrWhiteSpace(body))
        {
            return new DevLogBusPublishResult(1, false);
        }

        using var doc = JsonDocument.Parse(body);
        var published = doc.RootElement.TryGetProperty("published", out var value) && value.TryGetInt32(out var count)
            ? count
            : 1;
        return new DevLogBusPublishResult(published, false);
    }

    private static string TrimEndpoint(string? endpoint)
    {
        var value = string.IsNullOrWhiteSpace(endpoint) ? DefaultHttpEndpoint : endpoint.Trim();
        return value.TrimEnd('/');
    }
}

File: sdk/jvm/src/main/java/com/dansherwin/devlogbus/DevLogBusClient.java
package com.dansherwin.devlogbus;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public final class DevLogBusClient {
    public static final String DEFAULT_HTTP_ENDPOINT = "http://127.0.0.1:7423";
    public static final String REDACTED_VALUE = "[REDACTED]";

    private final String endpoint;
    private final String source;
    private final Duration timeout;
    private final RecordFilter filter;
    private final RecordRedactor redactor;
    private final HttpClient httpClient;

    public DevLogBusClient(Options options) {
        Options actual = options == null ? Options.builder().build() : options;
        this.endpoint = trimEndpoint(actual.endpoint);
        this.source = Objects.requireNonNullElse(actual.source, "");
        this.timeout = actual.timeout == null ? Duration.ofSeconds(2) : actual.timeout;
        this.filter = actual.filter;
        this.redactor = actual.redactor;
        this.httpClient = actual.httpClient == null ? HttpClient.newHttpClient() : actual.httpClient;
    }

    public PublishResult publishMessage(String level, String message, String attrsJson) throws IOException, InterruptedException {
        return publish(DevLogBusRecord.builder()
                .source(source)
                .level(level)
                .message(message)
                .attrsJson(attrsJson)
                .build());
    }

    public PublishResult publish(DevLogBusRecord input) throws IOException, InterruptedException {
        DevLogBusRecord record = input.source().isBlank()
                ? input.toBuilder().source(source).build()
                : input;
        if (filter != null && !filter.shouldPublish(record)) {
            return new PublishResult(0, true);
        }
        if (redactor != null) {
            record = redactor.redact(record);
        }

        HttpRequest request = HttpRequest.newBuilder(URI.create(endpoint + "/api/records"))
                .timeout(timeout)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(record.toJson()))
                .build();
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException("DevLogBus publish failed: HTTP " + response.statusCode());
        }
        return new PublishResult(1, false);
    }

    public DevLogBusLogger logger() {
        return new DevLogBusLogger(this);
    }

    public static RecordFilter dropSources(String... sources) {
        Set<String> blocked = new HashSet<>();
        for (String source : sources) {
            if (source != null && !source.isBlank()) {
                blocked.add(source.trim());
            }
        }
        return record -> !blocked.contains(record.source());
    }

    public static RecordRedactor redactMessage() {
        return record -> record.toBuilder().message(REDACTED_VALUE).build();
    }

    private static String trimEndpoint(String endpoint) {
        String value = Objects.requireNonNullElse(endpoint, DEFAULT_HTTP_ENDPOINT).trim();
        while (value.endsWith("/")) {
            value = value.substring(0, value.length() - 1);
        }
        return value.isEmpty() ? DEFAULT_HTTP_ENDPOINT : value;
    }

    public record PublishResult(int published, boolean filtered) {
    }

    public static final class Options {
        private String endpoint = DEFAULT_HTTP_ENDPOINT;
        private String source = "";
        private Duration timeout = Duration.ofSeconds(2);
        private RecordFilter filter;
        private RecordRedactor redactor;
        private HttpClient httpClient;

        private Options() {
        }

        public static Builder builder() {
            return new Builder();
        }

        public static final class Builder {
            private final Options options = new Options();

            public Builder endpoint(String endpoint) {
                options.endpoint = endpoint;
                return this;
            }

            public Builder source(String source) {
                options.source = source;
                return this;
            }

            public Builder timeout(Duration timeout) {
                options.timeout = timeout;
                return this;
            }

            public Builder filter(RecordFilter filter) {
                options.filter = filter;
                return this;
            }

            public Builder redactor(RecordRedactor redactor) {
                options.redactor = redactor;
                return this;
            }

            public Builder httpClient(HttpClient httpClient) {
                options.httpClient = httpClient;
                return this;
            }

            public Options build() {
                return options;
            }
        }
    }
}
